ls1intum/Artemis
# `Plagiarism checks`: Allow instructors to clean up and improve performance

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


Currently the plagiarism detection causes performance issues. This is due to some errors in the data model and the amount of plagiarism elements stored and used in this process.

### Description

Plagiarism Submission Elements now have a many-to-one relationship to plagiarism submission instead of a many-to-many, which was not necessary and avoids an additional table in the database. 

Instructors now have to option to clean up plagiarism runs. This means they can delete all plagiarism comparisons which don't have a status of DENIED or CONFIRMED, and they can also delete all plagiarism runs for an exercise if needed.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Exercise with very very similar submissions by at least two students

1. Log in to Artemis as an instructor (or editor)
2. Navigate to the plagiarism detection page for your exercise
3. Run the plagiarism detection
4. Check that the plagiarism comparisons are displayed correctly
5. Set some plagiarism comparisons to status DENIED or CONFIRMED
6. Open the clean up dialog
7. Confirm the clean up
8. Check that all plagiarism comparisons without a status of DENIED or CONFIRMED are removed
9. Open the clean up dialog again
10. Tick the box to delete all elements
11. Confirm the clean up
12. Check that no plagiarism result is shown
13. (only possible locally) Check that the plagiarism submission elements are also deleted

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| PlagiarismResource.java | 66% | 88% |
| plagiarism-inspector.component.ts | 76.47% | 90.32% |

### Screenshots



#### Button
![Bildschirmfoto 2022-06-02 um 00 17 10](https://user-images.githubusercontent.com/51987021/171511145-80c5fbf7-c85f-4929-9980-45148dae2c66.png)

#### Clean up dialog
![Bildschirmfoto 2022-06-02 um 00 17 16](https://user-images.githubusercontent.com/51987021/171511160-397cc57d-f005-4081-b709-37e3840805f6.png)

#### Clean up dialog - delete all
![Bildschirmfoto 2022-06-02 um 00 17 23](https://user-images.githubusercontent.com/51987021/171511183-f3fece00-bd60-4a40-930d-c70781eda9c6.png)
